### DISCUSSION TITLE
Create Declaration / Definition in current file

### DISCUSSION THREAD

**Discussion by FernandoGarcia** [microsoft/vscode-cpptools#10642] [General]
Hi!

I would to know how to create function prototype in an Arduino code while using Platformio.

I'm talking about this type of code:

main.cpp

```cpp
#include <Arduino.h>
#include <Bounce2.h>

#define NUM_RELAYS 3
#define INITIAL_STATE LOW  // Initial state = HIGH to active LOW relay model.

struct INFOS
{
  const byte relayPin;
  bool relayState;
  const byte buttonPin;
};

INFOS info[NUM_RELAYS] =
{
  {2, INITIAL_STATE, 18}, // Relay 0 pin, relay 0 state, button 0 pin
  {4, INITIAL_STATE, 19},
  {5, INITIAL_STATE, 21}
};

Bounce2::Button button[NUM_RELAYS] = {Bounce2::Button()};

void setup()
{
  Serial.begin(9600);

  for (byte i = 0; i < NUM_RELAYS; i++)
  {
    button[i].attach(info[i].buttonPin, INPUT_PULLUP);
    button[i].setPressedState(LOW);
    button[i].interval(5);

    pinMode(info[i].relayPin, OUTPUT);
    digitalWrite(info[i].relayPin, info[i].relayState);
  }
}

void loop()
{
  for (byte i = 0; i < NUM_RELAYS; i++)
  {
    button[i].update();

    if (button[i].pressed())
    {
      info[i].relayState = !info[i].relayState;
      digitalWrite(info[i].relayPin, info[i].relayState);
      printDebug(i);
      break;
    }
  }
}

void printDebug(byte i)
{
  Serial.print("Button ");
  Serial.print(i);
  Serial.print(" pressed, relayState ");
  Serial.print(i);
  Serial.print(": ");
  Serial.println(info[i].relayState);
}
```

The compiling error for this code is:

```
src/main.cpp: In function 'void loop()':
src/main.cpp:47:7: error: 'printDebug' was not declared in this scope
       printDebug(i);
       ^~~~~~~~~~
src/main.cpp:47:7: note: suggested alternative: 'Printable'
       printDebug(i);
       ^~~~~~~~~~
       Printable
*** [.pio/build/uno/src/main.cpp.o] Error 1
```

So I have always to create manually all functions prototypes on top of my code  in this way:

```cpp
void printDebug(byte i);
```

I have tried the [new feature](https://devblogs.microsoft.com/cppblog/vs-code-c-extension-january-update-create-definitions-and-declarations/) of C++ extension but it doesn't works well for this type of code.

The declaration is being placed in Arduino.h header file what is not correct for this case.

Thanks in advance.

Best regards.

**Comment by sean-mcmanus** [comment]
If Arduino.h does not exist in the current workspace, then it should be fixed in our next 1.15.0 release from https://github.com/microsoft/vscode-cpptools/issues/10402 .

Otherwise, there is no way to force it add the declaration to the current file, but you can add a `#include "main.h"` at the top of main.cpp to get it to add the declaration to main.h instead of another header.

**Comment by FernandoGarcia** [comment]
Hi!

Thanks for your reply!

The path to Arduino.h is `/home/fernandogarcia/.platformio/packages/framework-arduino-avr/cores/arduino/Arduino.h` and the path to my current workspace is: `/home/fernandogarcia/Documentos/PlatformIO/Projects/General tests`

I'll try including `main.h`.

Best regards.

**Comment by FernandoGarcia** [comment]
It doesn't worked.

![Captura de tela de 2023-03-08 17 31 33](https://user-images.githubusercontent.com/2778501/223843063-a9fcba9a-c388-4349-b8a5-bf29a84fb9bb.png)

**Reply by sean-mcmanus** [reply]
Did you save main.cpp?

**Reply by FernandoGarcia** [reply]
Saving the file or not the result is the same.

![Captura de tela de 2023-03-08 17 37 34](https://user-images.githubusercontent.com/2778501/223844228-5a6171a1-9689-4fb9-8f32-49327cb97e88.png)

**Reply by sean-mcmanus** [reply]
You may be hitting a bug related to https://github.com/microsoft/vscode-cpptools/issues/10619 -- if you add
```cpp
int iii;
int jjj;
```
to main.h, does that fix it? Otherwise, this might be some other bug we're not aware of yet, which might be fixed in 1.15.0 (since Arduino.h would be excluded), but I'm not sure.

**Reply by FernandoGarcia** [reply]
No adding this two variables doesn't solve the problem.

**Reply by sean-mcmanus** [reply]
Do the variables in main.h appear in the Outline view?
![image](https://user-images.githubusercontent.com/19859882/223904818-c644a1c4-cf83-45cf-b84b-58a7a10b2c0a.png)

If not that would indicate the database is in a bad state. Either way, it might be worth running C/C++: Reset IntelliSense Database in case it is a database issue.

**Reply by FernandoGarcia** [reply]
I think it's outlined.

![Captura de tela de 2023-03-09 00 03 30](https://user-images.githubusercontent.com/2778501/223905728-daf2c35f-47ff-47ff-bafa-3a54a700f3e1.png)

**Comment by FernandoGarcia** [comment]
Hi @sean-mcmanus !

I have installed the pre-release 1.15.0 and the problem remains.

Best regards.
